XMonad/Actions/MostRecentlyUsed.hs

```diff
@@ -116,7 +128,43 @@ mostRecentlyUsed
   -> KeySym   -- ^ The 'KeySym' corresponding to the key to which the action
               --   is bound.
   -> X ()
-mostRecentlyUsed mods tab = do
+mostRecentlyUsed mods key = do
+  (undo, toUndo) <- undoer
+  let undoably curThing withThing thing = curThing >>= \cur ->
+        when (cur /= thing) $ withThing thing >> toUndo (withThing cur)
+  withMostRecentlyUsed mods key $ \win Location{workspace,screen} ->
+    handlingRefresh $ do
+      undo
+      undoably curScreenId viewScreen screen
+      undoably curTag      greedyView workspace
+      mi <- gets (W.findTag win . windowset)
+      for_ mi $ \i -> do
+        undoably curTag greedyView i
+        mfw <- peek
+        for_ mfw $ \fw -> do
+          undoably (pure fw) focusWindow win
+  where
+    undoer :: (MonadIO m, Monoid a) => m (m a, m a -> m ())
+    undoer = do
+      ref <- io . newIORef $ pure mempty
+      let
+        undo = join (io $ readIORef ref) <* io (writeIORef ref $ pure mempty)
+        toUndo act = io $ modifyIORef ref (liftA2 (<>) act)
+      pure (undo, toUndo)
+    viewScreen :: ScreenId -> X Any
+    viewScreen scr = screenWorkspace scr >>= foldMap view
```

Comment:
Stupid question: do we actually need to do this here?  I don't think this is quite the same situation as https://github.com/xmonad/xmonad-contrib/pull/543.  We should already be doing the "correct thing" here; that is, not generating windowsets but operating on the current windowset instead, as well as (and this is important) declaring ourselves busy and not inserting new windows into the history in the logHook when cycling.  All state should be preseved in this way, with no ill effects.

Reply:
Not really a stupid question; this is more a matter of UX than necessity. I'm considering the intermediate actions to be *previews* while the final action is the *selection*—and a preview that isn't selected should be erased as if it never happened.

I didn't think about this too deeply in the previous draft because I figured `greedyView` would take care of it, but that's not quite right either. With the screen tracking version, it's more obvious that a series of unmitigated previews will leave unrelated screens focused on historic windows we only wanted to tab past. 

What I'd really like to write here is much simpler:
```haskell
mostRecentlyUsed mods key = withWindowSet $ \ws ->
  withMostRecentlyUsed mods key $ \win Location{workspace,screen} ->
    handlingRefresh . void $ do
      modifyWindowSet (const ws)
      screenWorkspace screen >>= foldMap view
      greedyView workspace
      focusWindow win
```
But there, I gather, we really would run into the same problem as #543 when we discard the layout updates performed in each refresh. Hence, I had to do something more complex: build and discharge an "inverse" alongside each preview.

Reply:
> Not really a stupid question; this is more a matter of UX than necessity. I'm considering the intermediate actions to be _previews_ while the final action is the _selection_—and a preview that isn't selected should be erased as if it never happened.
>
> I didn't think about this too deeply in the previous draft because I figured `greedyView` would take care of it, but that's not quite right either. With the screen tracking version, it's more obvious that a series of unmitigated previews will leave unrelated screens focused on historic windows we only wanted to tab past.

Ah, I see; that would explain the use of `PureX` as well.  It's true that this is nicer conceptually.

> What I'd really like to write here is much simpler:
>
> ```haskell
> mostRecentlyUsed mods key = withWindowSet $ \ws ->
>   withMostRecentlyUsed mods key $ \win Location{workspace,screen} ->
>     handlingRefresh . void $ do
>       modifyWindowSet (const ws)
>       screenWorkspace screen >>= foldMap view
>       greedyView workspace
>       focusWindow win
> ```
>
> But there, I gather, we really would run into the same problem as #543 when we discard the layout updates performed in each refresh. Hence, I had to do something more complex: build and discharge an "inverse" alongside each preview.

Isn't it more like we queue up the state mutations, rather than discard them?  If I change the definition of `mostRecentlyUsed` to something like[^1]

``` haskell
mostRecentlyUsed mods key =
  withMostRecentlyUsed mods key $ \win Location{workspace,screen} ->
    handlingRefresh . void $ do
      screenWorkspace screen >>= foldMap view
      greedyView workspace
      focusWindow win
```

then for example X.L.ShowWName (the layout that gave us headaches in #543) seems to work correctly (i.e., correctly update its state).  While it's true that the small popup window that it spawns stays open while cycling windows on the same workspace, it correctly vanishes when either switching workspaces, or once one lets go of the relevant keybinding.

[^1]: I deleted `modifyWindowSet (const ws)` because I'm not quite sure why this is needed here.  Admittedly I'm not well-versed with `PureX`.

Reply:
We're not quite on the same page here; let me rewrite that without `PureX`, which I've actually only used because it's *cleaner*. Though I'll warn you, this one isn't type checked:

```haskell
mostRecentlyUsed mods key = withWindowSet $ \ws ->
  withMostRecentlyUsed mods key $ \win Location{workspace,screen} -> windows
    $ W.focusWindow win 
    . W.greedyView workspace
    . maybe id W.view (W.lookupWorkspace screen)
    . const ws
```

The `modifyWindowSet (const ws)`, now `. const ws`, is precisely where we discard prior changes—both unwanted previews and wanted layout updates—it's the key to the simplified but bug-inducing version. I certainly expect the version without it to work without bugs, but also without the UX I talked about.

Reply:
Sorry, I still don't quite get it; maybe I'm not seeing the whole picture here.  To me, `withWindowSet $ \ws -> windows (const ws)` is equivalent to `windows id`; it certainly shouldn't discard any kind of state.  What am I missing?

Reply:
The windowset `ws` is grabbed only once at the very beginning, while the function `\win Location{...} -> windows $ ... . const ws`, runs repeatedly, once for each time `key` is tapped. The first time it runs, it will indeed be a no-op. The second time, however, it will erase the changes left behind by the first invocation, and so on.

Reply:
Aaaah, right, that make sense.